emitter.ts

```diff
@@ -131,258 +123,218 @@ export async function emitter(
-          finalizationQueue.map((evt) =>
-            `${evt.blockNumber}-${evt.logIndex} (${
-              formatDate(
-                new Date(Number(evt.blockTimestamp) * 1000),
-                "yyyy-MM-dd HH:mm:ss",
-              )
-            })`
-          )
-            .join(", ")
-        }.`
-        : "No events left in finalization queue."
-    );
-  }
+            };
+            const newLockedTimestamp = new Date();
+            if (
+              (await prisma.event.updateMany({
+                where: { ...where, lockedTimestamp: x.lockedTimestamp },
+                data: { lockedTimestamp: newLockedTimestamp },
+              })).count <= 0
+            ) {
+              logger.warning(
+                `Lock was modified by another party, aborting emit, blockNumber: ${x.blockNumber}  logIndex: ${x.logIndex}.`,
+              );
+              return undefined;
+            }
+            const nextRetryUrls = (await Promise.all(x.url.map(async (url) => {
+              logger.info(
+                `${
+                  x.retry ? "Retry p" : "P"
+                }osting event destination: ${url}  blockNumber: ${x.blockNumber}  logIndex: ${x.logIndex}.`,
+              );
+              try {
+                const response = await fetch(url, {
+                  method: "POST",
+                  headers: { "Content-Type": "application/json" },
+                  body: losslessJsonStringify(serializeEventResponse(x)),
+                });
+                if (response.ok) return undefined;
+                logger.error(
+                  `Event emit request failed with an HTTP error, status: ${response.status}  blockNumber: ${x.blockNumber}  logIndex: ${x.logIndex}  url: ${url}  response body: ${response.text()}`,
+                );
+              } catch (e) {
+                logger.error(
+                  `Event emit request failed with an exception, blockNumber: ${x.blockNumber}  logIndex: ${x.logIndex}  url: ${url}: ${
+                    e.stack ?? e.message
+                  }`,
+                );
+                return url;
+              }
+            }))).filter((x) => x != undefined) as string[];
+            await prisma.failedUrl.deleteMany({ where });
+            await Promise.all(
+              nextRetryUrls.map((url) =>
+                prisma.failedUrl.create({ data: { ...where, url } })
+              ),
+            );
+            if (nextRetryUrls.length <= 0) {
+              logger.info(
+                `Event post success for all webhook URLs, blockNumber: ${x.blockNumber}  logIndex: ${x.logIndex}.`,
+              );
 
-  const abortController = new AbortController();
-  const runningPromise = block(abortController.signal);
+              await prisma.event.updateMany({
+                where,
+                data: {
+                  lockedTimestamp: null,
+                  emittedTimestamp: new Date(),
+                },
+              });
```

Comment:
I think it should already ensure atomicity because the other queries that writes to the lock specifies the lock that the code intends to change. Besides, I believe that transactions can ensure order or atomicity inside the transaction itself, but doesn't ensure the order between different transactions (though I might be wrong 😇).